### PR TITLE
Dev guide for WHERE predicate pushdown

### PR DESCRIPTION
**Pull request by MiguelWeezardo**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by MiguelWeezardo** [comment]
Original PR: https://github.com/nineinchnick/trino/pull/9

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:411]: If this doesn't explain all the nuances, only the one example below, it's not relevant. Can you rephrase this to be more explicit about having to ensure that if a constraint gets pushed down, it should work exactly the same in the underlying data source and produce the same results?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:418]: This is confusing. I'm not sure how I'd use this information when writing a non-JDBC connector. Note that I think we should have a whole separate page about JDBC connectors.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:422]: This is too passive, who handles the `TupleDomains`? Where do the expressions get split? Does this need to happen in the connector?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:255]: This paragraph duplicates the next one and the previous one combined. What are you trying to say here that's not already in the surrounding paragraphs?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:312]:
```suggestion
When the query plan contains a ``Filter`` operation, Trino
```

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:322]: This is true for all `applyXXX` functions, so maybe it should be added earlier, somewhere around line 250?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:327]: Connectors push down the `Filter` operation, not the `WHERE` clause.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:329]:
```suggestion
constraint has already been applied.
```

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:352]: This guide should not refer to any specific connectors. What does `immediate filtering` mean? `applyFilter` doesn't handle any data, other services do, like the split manager and record set provider (or page source provider).

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:357]: How is that important when developing a new connector? If a connector could only handle either the summary or predicates, can this be stated explicitly?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:364]:
```suggestion
In this case the ``y > 5`` part of the filter will be applied Trino,
and not pushed down.
```

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/connectors.rst:360]: Connectors don't receive queries. They are used in catalogs which can be referenced in queries.
```suggestion
If a constraint can only be pushed down partially, e.g. a connector for
a database that doesn't support range matching is used in a query with
```

File: docs/src/main/sphinx/develop/connectors.rst
Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:352]: By `immediate filtering` I meant filtering that occurs inside `applyFilter`, since we can't store the predicate in a `TableHandle`. How this would be done depends on the DB implementation - in Hive connector we query the HMS for partitions, filter them using the predicate, and store the resulting list in `HiveTableHandle`. I guess sharded DBs could use a similar approach if their API allows querying for a list of shard keys.

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:345]: This is an implementation detail that we shouldn't include in the guide. I'm not entirely sure if `applyFilter` should be doing time-consuming operations like talking to external services, since query planning should be done as quickly as possible. Instead, maybe it's better to do it in the split manager, using the table handle returned here.

It's enough to say that predicate objects cannot be reused outside of this method, we don't need to go into details about how to deal with it.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/connectors.rst:350]: The API ambiguity should be explained by git commits or in the API Javadocs.

Regarding the preference, it should be possible only to handle one kind of constraint, either the summary or predicate, not both. So instead of writing (pseudo) code like:
```
if (predicate) {
 // handle predicate
} else {
 // handle summary
}
```
I'd rather write:
```
if (summary) {
  // handle
}
```
Are there cases when a predicate is present but the summary is not? Is that a bug?

Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:350]: I think there can be cases where the summary is present, but there is no predicate, but not the other way around.

Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:350]: Also, handling of the predicate is entirely optional - it's correct (but less efficient) to ignore it and focus on summary.

**Comment by mosabua** [comment]
There is a lot of feedback ... let me know when that is addressed and then I can look... sounds good?

In general.. shorter code snippets and less of them directly in the docs is probably better ..

**Comment by MiguelWeezardo** [comment]
@mosabua You can take a look now, thanks!

File: docs/src/main/sphinx/develop/connectors.rst
Comment (mosabua) [inline docs/src/main/sphinx/develop/connectors.rst:525]:
```suggestion
``remainingExpression`` and are evaluated by the Trino engine.
```

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/develop/connectors.rst
Comment (MiguelWeezardo) [reply docs/src/main/sphinx/develop/connectors.rst:306]: It's always possible for the `ScanFilterProject` operation to be optimized out at a higher level, so I think the best we can do is "will likely contain".

**Comment by MiguelWeezardo** [comment]
@mosabua Could you take another look?

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:311]: No use of "would"

...= 3`` translates into

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:312]: - No use of "would"
- Do not use "say" - google style guide requires plain, non-idiomatic language, and inanimate objects don't actually say things. 
 
- You can use "means" instead of "would say".

Use "must be" instead of "has to be"

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:362]: for example when a

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:366]: In this case, the ``y > 5`` condition is

("part of the filter" is very informal. Please be specific about that part of the filter. It is a condition.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:407]: Far too informal. This is blog post/user guide language, not reference material language. the material in the parenthesis does not make sense and in general, parenthesis should be avoided at all costs.. "Not exactly represented" is confusing phrasing.

Suggest:

The following, more complex example shows data types from Trino that are available directly in the underlying data source, and must be mapped:

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:481]: Writing is too informal/blog-posty.

This example illustrates implementing a base class for many JDBC connectors while handling the specific requirements of multiple JDBC-compliant data sources. It ensure that if a constraint gets pushed down, it works
exactly the same in the underlying data source, and produce the same results
as it would in Trino.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:487]: Move this to previous paragraph. 

Do not use "we" - too informal.

For example, in databases where string comparisons are is case-insensitive, pushdown does not work, as string
comparison operations in Trino are case-sensitive.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:500]: Do not use "we" - too informal.
Do not use "usually" (or "could"  or "would") as per google dev style guide. In reference material, writing must be definite.
Do not use directions such as above, below. User following/preceding (g dev style guide)

Also, I have used "interface" below, because you seem to call it an interface in the paragraph. Please ensure that it is being named *precisely* what it is (interface/class/method, etc).

The ``PredicatePushdownController`` interface determines if a column domain can be pushed down in JDBC-compliant data sources. In the preceding example, it is called from a ``JdbcClient`` implementation specific to that database. In non-JDBC-compliant data sources, type-based push downs are implemented directly, without going through the ``PredicatePushdownController`` interface.

Now let's add expression pushdown to our code, enabled by a session flag:

**Review by m57lyra — COMMENTED** [review]
Writing style is too informal (blog post-y), and there are style guide issues (for inclusivity and grammar).

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:311]: just constraint, no initial cap and no double backticks

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:319]: and returns either 

(no should - it is against developer style guide)

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:324]: Connectors must return

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:325]: cannot

(no contractions - against developer style guide)

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:326]: invocation, even

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:327]: Connectors must also return

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:343]: A constraint contains the following elements:

* A ``TupleDomain`` defining the mapping between columns and their domains. A ``Domain`` is either a list of possible values, or a list of ranges, and also contains information about nullability.
* Expression for pushing down function calls.
* Map of assignments from variables in the expression to columns.
* (optional) Predicate which tests a map of columns and their values;
  it cannot be held on to after the ``applyFilter`` call returns.
* (optional) Set of columns the predicate depends on; must be present
  if predicate is present.

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:349]: It is not possible to

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:362]: partially pushed down

File: docs/src/main/sphinx/develop/connectors.rst
Comment (m57lyra) [inline docs/src/main/sphinx/develop/connectors.rst:363]: does not

**Review by m57lyra — COMMENTED** [review]
Looking better, but still some things to fix

**Review by m57lyra — APPROVED** [review]
LGTM!

**Comment by m57lyra** [comment]
@MiguelWeezardo - Did you actually make additional changes after my approval that i need to review or is this just the squash and merge?

**Comment by MiguelWeezardo** [comment]
> @MiguelWeezardo - Did you actually make additional changes after my approval that i need to review or is this just the squash and merge?

It's just a squash.

**Comment by MiguelWeezardo** [comment]
@m57lyra Should I involve a maintainer to get this merged? If so, I'll probably need an official OK from you.

**Comment by MiguelWeezardo** [comment]
@hashhar I'd appreciate if you took a look at this and possibly merge if you think this is useful.

**Comment by hashhar** [comment]
Sure, will take a look. Some quick comments follow.

**Comment by MiguelWeezardo** [comment]
@hashhar Could you take another look and merge if OK?

**Comment by MiguelWeezardo** [comment]
@hashar Could you take another look and merge if OK?

**Review by raunaqmorarka — APPROVED** [review]

**Pull request merged**
